vulnerabilities/views.py

```diff
@@ -19,50 +19,118 @@
             except PageNotAnInteger:
                 packages = Paginator(packages, 50).get_page(1)
-            packages = Paginator(packages, 50).get_page(page_no)
+            packages = pages.get_page(page_no)
+
             context["packages"] = packages
             context["searched_for"] = urlencode(
                 {param: request.GET[param] for param in request.GET if param != "page"}
             )
             context["result_size"] = result_size
 
+            if len(request.GET["type"]):
+                package_type = request.GET["type"]
+                context["package_type"] = package_type
+
+            if len(request.GET["name"]):
+                package_name = request.GET["name"]
+                context["package_name"] = package_name
+
+            if result_size == 0:
+                context = {
+                    "package_search": "The VCIO DB does not contain a record of the package you entered -- "
+                    + request.GET["name"]
+                    + ".",
+                    "debug_ui": settings.DEBUG_UI,
+                }
+
+                if request.GET.get("template") == "packages":
+                    context["package_form"] = PackageForm(request.GET or None)
+                    context["template_name"] = "packages.html"
+                    return render(request, "packages.html", context)
+                elif request.GET.get("template") == "package_details":
+                    context["package_form"] = PackageForm(request.GET or None)
+                    context["template_name"] = "package_update.html"
+                    return render(request, "package_update.html", context)
+                elif request.GET.get("template") == "index":
+                    context["package_form"] = PackageForm(request.GET or None)
+                    context["vuln_form"] = CVEForm(request.GET or None)
+                    context["template_name"] = "index.html"
+                    return render(request, "index.html", context)
+                else:
+                    context["package_form"] = PackageForm(request.GET or None)
+                    context["vuln_form"] = CVEForm(request.GET or None)
+                    context["template_name"] = "index.html"
+                    return render(request, "index.html", context)
+
+        context["package_form"] = PackageForm(request.GET or None)
+        context["template_name"] = self.template_name
         return render(request, self.template_name, context)
 
     @staticmethod
     def request_to_queryset(request):
         package_type = ""
         package_name = ""
+        purl = ""
 
         if len(request.GET["type"]):
             package_type = request.GET["type"]
 
         if len(request.GET["name"]):
             package_name = request.GET["name"]
 
+        # Check whether the input value is a syntactically-correct purl
+        try:
+            purl = PackageURL.from_string(package_name)
+            return list(
```

Comment:
The QS should live outside the try block.

Reply:
@tdruez I don't understand -- what's the reason to move the Qs outside the `try` block, and how would that be done?

Reply:
Perhaps you're suggesting that I move the entire `return()` into a new method, leaving only `purl = PackageURL.from_string(package_name)`, which tests whether the input is a syntactically-correct purl?  Pass `purl` to the new method, keeping the `try` focused on its sole purpose: the `purl` test?

Reply:
> what's the reason to move the Qs outside the try block

In general, you want to keep only one piece of logic that could fail within a `try` block.
An exception raised by converting a PURL has nothing to do with an exception filtering a QuerySet.

Also, using generic `except:` is bad practice. You should always try to be explicit about the exception you want to catch. Since you want to "Check whether the input value is a syntactically-correct purl", let's catch that specific Exception:
```
>>> PackageURL.from_string('wrong syntax')
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/Volumes/nexB/repos/scancode.io/lib/python3.9/site-packages/packageurl/__init__.py", line 354, in from_string
    raise ValueError(
ValueError: purl is missing the required "pkg" scheme component: {repr(purl)}.
```

The packageurl raise a `ValueError` in that case, that's the exception we want to catch.
You code could become:

```
try:
    purl = PackageURL.from_string(package_name)
except ValueError:
    purl = None

if purl:
    packages = ...
else:
    packages = ...

return list(packages)
```

This way, this are properly grouped, the code is more readable:
1. we deal with the purl
2. we prepare a queryset based on the available data
3. we return the fetched objects

> keeping the try focused on its sole purpose: the purl test?

Yes, bottom line is to always try to keep you `try` block as focused as possible.

Reply:
Far out.  Thank you for the detailed explanation @tdruez .  I'll tackle this in the A.M.  👍

Reply:
See new issue https://github.com/nexB/vulnerablecode/issues/876